qcodes/sphinx_extensions/add_parameter.py

```diff
@@ -0,0 +1,268 @@
+    if modify_dosctring and hasattr(obj, _ADD_PARAMETER_ATTR_NAME):
+        signature = "()"
+    return (signature, return_annotation)
+
+# ######################################################################################
+# Setup sphinx extension
+# ######################################################################################
+
+def dont_skip_qcodes_params(app, what, name, obj, skip, options) -> bool:
+    """
+    Intercepts :code:`"autodoc-process-signature"` and sets the signature to
+    :code:`"()"`.
+    """
+    if hasattr(obj, _ADD_PARAMETER_ATTR_NAME):
+        return not app.config["qcodes_parameters_force_documenting"]
+    return skip
+
+def setup(app):
+    """Called by sphinx to setup the extension."""
+    app.setup_extension('sphinx.ext.autodoc')  # Require autodoc extension
+
+    # Register the extension configuration parameters
+    app.add_config_value(
+        name="qcodes_parameters_spec_in_docstring",
+        default=True,
+        rebuild="html",
+        types=[bool]
+    )
+    app.add_config_value(
+        name="qcodes_parameters_spec_with_links",
+        default=True,
+        rebuild="html",
+        types=[bool]
+    )
+    app.add_config_value(
+        name="qcodes_parameters_force_documenting",
+        default=True,
+        rebuild="html",
+        types=[bool]
+    )
+    # we can't have a method in the class with the same name as the desired
+    # parameter, therefore we patch the method name displayed in the sphinx docs
+    # monkey patching `MethodDocumenter`
+    # e.g. `_parameter_time` -> `_parameter_time` when displayed in the docs
+    autodoc.MethodDocumenter.format_name = format_name
+
+    # Patch several parts of autodoc and sphinx so that we have a nice
+    # "parameter " prefix added in the docs output, similar to, e.g., "classmethod ".
+    # Maybe a custom Documenter inheriting from MethodDocumenter could be developed
+    # though it is not clear if that would allow for all the functionality we want
+    autodoc.MethodDocumenter.add_directive_header = add_directive_header
+    sphinx_domains_python.PyMethod.option_spec.update(
+        {"parameter": sphinx_domains_python.PyMethod.option_spec["classmethod"]}
+    )
+    sphinx_domains_python.PyMethod.get_signature_prefix = get_signature_prefix
+
+    # enforce always documenting the decorated private methods whose name is
+    # prefixed with _ADD_PARAMETER_ATTR_NAME
+    app.connect("autodoc-skip-member", dont_skip_qcodes_params)
+    # convert the signature of the decorated methods to a nicely formatted list in the
+    # docstring
+    app.connect("autodoc-process-docstring", add_parameter_spec_to_docstring)
+    app.connect("autodoc-process-signature", clear_parameter_method_signature)
+
+    return {
+        'version': '0.1',
+        'parallel_read_safe': True,  # Not tested, should not be an issue
+        'parallel_write_safe': True,  # Not tested, should not be an issue
```

Comment:
I do not want to limit anyone that might be using the parallel sphinx build. Let's fix it if it becomes a problem. (I cannot test because on mac parallel build is not supported at all.)